### ISSUE TITLE
WebGL Support

### ISSUE THREAD

**Issue by cwfitzgerald** [BVE-Reborn/rend3#18]
Currently, rend3 uses the `PUSH_CONSTANT | MAPPABLE_PRIMARY_BUFFERS` features which are incompatible with the web. If people want rend3 to run on the web, we need to allow these features to be unsupported, ideally without removing the ability to use them on native.

While I don't particularly have much interest in web support, if someone else does, I'd be happy to support it.

**Comment by cwfitzgerald** [comment]
Now with the new render routines, rend3 is compatible with WebGPU, but not WebGL.

**Comment by cwfitzgerald** [comment]
Considering that WebGPU is 4 months away from being stable in chrome, and WebGL would be quite a lot of work, I'm going to close this issue in favor of getting things working on WebGPU.

**Comment by daxpedda** [comment]
Considering that Firefox and Safari didn't signal WebGPU stabilization, I will probably be interested in contributing WebGL support at some point in the future (couple of month or so).

If WebGPU is still not in sight for both browsers by then, would you still be interested in contributions for WebGL support?

**Comment by cwfitzgerald** [comment]
So chrome is a lot further along than the rest, though they have moved the target for stabilization back to Q3 next year. That being said, rend3 runs fine on the WebGPU origin trial on stable chrome now that #260 as merged.

I'm not sure if WebGL is really possible at this point. It would require an even-more-downlevel path that doesn't use storage buffers, compute shaders, and the like. 

The direction I was planning to go with rend3 was to assume a device that supports at least baseline WebGPU and unify the two current paths so we can have good performance on the web as well. I think the thing that makes rend3 unique is the use of modern rendering features to get much better performance than was otherwise possible.

tl;dr: WebGPU is still close, and supporting <WebGPU would require bending over backwards, when there are other libraries that can fill that role.

**Comment by daxpedda** [comment]
> when there are other libraries that can fill that role

Good 3d rendering libraries are hard to come by!

I understand your reasoning though and I think the decision to embrace modern available features to improve the performance is great! Too bad for WebGL really 😄.

**Comment by emilk** [comment]
As I understand it, rend3 uses `wgpu` which has a WebGL backend, but I'm not sure if there are some limitations to using the WebGL backend of wgpu. Maybe someone else has more knowledge about this?

**Comment by cwfitzgerald** [comment]
@emilk the main things you lose are storage buffers, compute shaders, and indirect drawing, all limitations of WebGL2/GLES3.0 itself. 

These are all things that I want to keep using in rend3, so I decided to not have webgl. If you don't need those features, wgpu's WebGL2 support is fine.

**Comment by emilk** [comment]
Thanks for the clarification!
